template/cordova/lib/package.js

```diff
@@ -113,16 +112,32 @@ function getPackageName(platformPath) {
 
 // returns one of available devices which name match with provided string
 // return rejected promise if device with name specified not found
-module.exports.findDevice = function (target) {
+module.exports.findDevice = function (deploymentTool, target) {
     target = target.toLowerCase();
-    return module.exports.listDevices().then(function(deviceList) {
+    return deploymentTool.enumerateDevices().then(function(deviceList) {
         // CB-7617 since we use partial match shorter names should go first,
         // example case is ['Emulator 8.1 WVGA 4 inch 512MB', 'Emulator 8.1 WVGA 4 inch']
-        var sortedList = deviceList.concat().sort(function (l, r) { return l.length > r.length; });
-        for (var idx in sortedList){
-            if (sortedList[idx].toLowerCase().indexOf(target) > -1) {
-                // we should return index based on original list
-                return Q.resolve(deviceList.indexOf(sortedList[idx]));
+        // In CB-9283, we need to differentiate between emulator, device, and target.
+        // So, for emulators to honor the above CB-7617, we preserve the original behavior.
+        // Else, we choose either the target by ID (DeviceInfo.index) or if it's just device,
+        // we choose the default (aka first) device.
+        if (target === 'emulator') {
+            var sortedList = deviceList.concat().sort(function (l, r) { return l.toString().length > r.toString().length; });
+            for (var idx in sortedList){
```

Comment:
If I understand correctly, `sortedList` is an array and foreach is quite a bad practice for arrays - unless we really have a case of a sparse array.